base/nullable.jl

```diff
@@ -78,3 +78,61 @@ function hash(x::Nullable, h::UInt)
         return hash(x.value, h + nullablehash_seed)
     end
 end
+
+# as collection
+length(x::Nullable) = x.isnull ? 0 : 1
+endof(x::Nullable)  = length(x)
+
+# indexing is either without index, or with 1 as index
+# generalized linear indexing is not supported
+linearindexing{T}(::Nullable{T}) = LinearFast()
+function getindex(x::Nullable)
+    @boundscheck x.isnull && throw(NullException())
+    x.value
+end
+function getindex(x::Nullable, i::Integer)
+    @boundscheck (x.isnull | (i ≠ 1)) && throw(BoundsError(i, x))
+    x.value
+end
+
+# convenience method for getindex without bounds checking
+unsafe_getindex(x::Nullable) = x.value
+unsafe_getindex(x::Number) = x
+
+# convenience method for detecting null value
+isnullvalue(x::Nullable) = x.isnull
+isnullvalue(x::Number) = false
+
+# iteration protocol
+start(x::Nullable) = 1
+next(x::Nullable, i::Integer) = x.value, 0
+done(x::Nullable, i::Integer) = x.isnull | (i == 0)
+
+# higher-order functions
+function filter{T}(p, x::Nullable{T})
+    if x.isnull
+        x
+    elseif p(x.value)
+        x
+    else
+        Nullable{T}()
+    end
+end
+
+function map{T}(f, x::Nullable{T})
+    if x.isnull
+        Nullable{Union{}}()  # Union{} for consistency with arrays
+    else
+        Nullable(f(x.value))
+    end
+end
+
+function map(f, xs::Nullable...)
+    if all(isnull, xs)
+        Nullable()
+    elseif !any(isnull, xs)
+        Nullable(map(f, map(unsafe_getindex, xs)...))
+    else
+        throw(DimensionMismatch("expected all null or all nonnull"))
```

Comment:
I would tend to think like @hayd on this one. In Haskell, `zipWith` on  lists has no problem with one list shorter than the other, it makes a lot of sense and I don't see why our map don't act similarly, both for arrays and nullables. Also, when lifting via map an operation on two or more nullables, I would naturally interpret getting a isnull back as an "error": the computation couldn't be done because of missing values. Having an error thrown when the inputs are not all isnull is redundant, as you have now to different ways to be notified of a failed computation, and you have to check both for an isnull result and for a possible thrown exception. `Maybe` in Haskell can be used to encode an error at the type level, and lifting `f` on `Maybe` would be naturally implemented as
```
map2 f mx my = do
    x <- mx
    y <- my
    return $ f x y
```
which returns `Nothing` if one or more inputs are Nothing (that said, applying `map2` on lists doesn't correspond to our multiple-list-arguments map).